### PR TITLE
Employee Service + Resource + Filter --> adding JWT cookie include: employeeID, companyID, Role

### PR DESCRIPTION
**Pull request by RonAzar**
Add logout functionality, create JWT for login, and secure session management with cookies

### CODE REVIEW

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/EmployeeAuthResource.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import com.hibob.academy.employee_feedback_feature.service.EmployeeAuthenticationService
+import com.hibob.academy.filters.AuthenticationFilter.Companion.COOKIE_NAME
+import jakarta.ws.rs.*
+import jakarta.ws.rs.core.MediaType
+import jakarta.ws.rs.core.NewCookie
+import jakarta.ws.rs.core.Response
+import org.springframework.stereotype.Controller
+
+@Controller
+@Produces
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/EmployeeAuthResource.kt:13]: why are you using produces?

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/service/EmployeeAuthenticationService.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import io.jsonwebtoken.Jwts
+import io.jsonwebtoken.SignatureAlgorithm
+import jakarta.ws.rs.core.NewCookie
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+import java.util.*
+
+@Component
+class EmployeeAuthenticationService @Autowired constructor(private val employeeDao: EmployeeDao) {
+    companion object{
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/service/EmployeeAuthenticationService.kt:16]: remove this if you are not using the companion

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/service/EmployeeAuthenticationService.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    fun createJwtTokenForEmployee(employeeLogin: EmployeeLogin): String {
+        val authenticatedEmployee = employeeDao.authenticateEmployee(employeeLogin)
+        return Jwts.builder()
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/service/EmployeeAuthenticationService.kt:21]: new line before return

File: src/main/kotlin/com/hibob/academy/filters/AuthenticationFilter.kt [hunk 4]
```diff
@@ -15,30 +13,39 @@ import jakarta.ws.rs.ext.Provider
...
 
     override fun filter(requestContext: ContainerRequestContext) {
         if (requestContext.uriInfo.path == LOGIN_PATH) return
 
         val cookie = requestContext.cookies[COOKIE_NAME]?.value
 
-        if (!verify(cookie)) {
-            requestContext.abortWith(Response.status(Response.Status.UNAUTHORIZED).entity("Invalid or expired token").build())
+        val claims = verifyAndExtractClaims(cookie)
+        if (claims == null) {
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/filters/AuthenticationFilter.kt:27]: new line before if

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/EmployeeAuthResource.kt (hunk 1 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/employee_feedback_feature/resource/EmployeeAuthResource.kt:13]: Oops my bad...

**Review by vasofi — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/hibob/academy/filters/AuthenticationFilter.kt [hunk 5]
```diff
@@ -15,30 +13,39 @@ import jakarta.ws.rs.ext.Provider
...
+            requestContext.abortWith(
+                Response.status(Response.Status.UNAUTHORIZED).entity("Invalid or expired token").build()
+            )
+            return
```
[the remark this answers is not in the document]
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/filters/AuthenticationFilter.kt:33]: Fixed

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/dao/EmployeeDao.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        )
+    }
+
+    fun authenticateEmployee(loginEmployee: EmployeeLogin): EmployeeData {
+        return sql.select(employee.id, employee.firstName, employee.lastName, employee.role, employee.companyId)
+            .from(employee)
+            .where(employee.companyId.eq(loginEmployee.companyId))
+            .and(employee.firstName.eq(loginEmployee.firstName))
+            .and(employee.lastName.eq(loginEmployee.lastName))
+            .fetchOne(employeeDataMapper) ?: throw IllegalArgumentException("User not found: Incorrect details provided")
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/dao/EmployeeDao.kt:27]: for general knowledge ( I know you were taught differently, so no need to change) - we want our DAOs to be "dumb".
The dao doesn't need to throw an exception, we can handle it later, maybe in some cases it shouldn't be an exception? the service can decide

File: src/main/kotlin/com/hibob/academy/employee_feedback_feature/dao/EmployeeDao.kt [hunk 7]
```diff
@@ -8,6 +8,7 @@ import org.springframework.stereotype.Component
 
 @Component
 class EmployeeDao @Autowired constructor(private val sql: DSLContext) {
+
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/employee_feedback_feature/dao/EmployeeDao.kt:11]: remove new line

File: src/main/kotlin/com/hibob/academy/filters/AuthenticationFilter.kt [hunk 8]
```diff
@@ -15,30 +13,39 @@ import jakarta.ws.rs.ext.Provider
 class AuthenticationFilter : ContainerRequestFilter {
```
Comment (vasofi) [inline src/main/kotlin/com/hibob/academy/filters/AuthenticationFilter.kt:14]: remove ne line at the top of the class

**Review by vasofi — APPROVED** [review]

**Pull request merged**